File: SmartReceipts/Persistence/Migrations/DatabaseUpgradeToVersion12.m
//
//  DatabaseUpgradeToVersion12.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 26/04/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import "DatabaseUpgradeToVersion12.h"
#import "DatabaseTableNames.h"
#import "Database.h"
#import "Database+PaymentMethods.h"
#import "Database+Functions.h"

@implementation DatabaseUpgradeToVersion12

- (NSUInteger)version {
    return 12;
}

- (BOOL)migrate:(Database *)database {
    [AnalyticsManager.sharedManager recordWithEvent:[Event startDatabaseUpgrade:self.version]];
    
    NSArray *alterTrips = @[@"ALTER TABLE ", TripsTable.TABLE_NAME, @" ADD ", TripsTable.COLUMN_FILTERS, @" TEXT"];
    NSArray *alterReceipts = @[@"ALTER TABLE ", ReceiptsTable.TABLE_NAME, @" ADD ", ReceiptsTable.COLUMN_PAYMENT_METHOD_ID, @" INTEGER REFERENCES ", PaymentMethodsTable.TABLE_NAME, @" ON DELETE NO ACTION"];

    BOOL result = [database createPaymentMethodsTable]
            && [database executeUpdateWithStatementComponents:alterTrips]
            && [database executeUpdateWithStatementComponents:alterReceipts];
    
    [AnalyticsManager.sharedManager recordWithEvent:[Event finishDatabaseUpgrade:self.version success:result]];
    
    return result;
}

@end

File: SmartReceipts/Persistence/Migrations/DatabaseUpgradeToVersion12.swift
//
//  DatabaseUpgradeToVersion12.swift
//  SmartReceipts
//
//  Created by Bogdan Evsenev on 30/03/2019.
//  Copyright © 2019 Will Baumann. All rights reserved.
//

import Foundation

class DatabaseUpgradeToVersion12: DatabaseMigration {
    
    var version: Int {
        return 12
    }
    
    func migrate(_ database: Database) -> Bool {
        AnalyticsManager.sharedManager.record(event: Event.startDatabaseUpgrade(version))
        
        let alterTrips = ["ALTER TABLE ", TripsTable.Name, " ADD ", TripsTable.Column.Filters, " TEXT"]
        let alterReceipts = ["ALTER TABLE ", ReceiptsTable.Name, " ADD ", ReceiptsTable.Column.PaymentMethodId, " INTEGER REFERENCES ", PaymentMethodsTable.Name, " ON DELETE NO ACTION" ]
        
        let result = database.createPaymentMethodsTable()
            && database.executeUpdate(withStatementComponents: alterTrips)
            && database.executeUpdate(withStatementComponents: alterReceipts)
        
        AnalyticsManager.sharedManager.record(event: Event.finishDatabaseUpgrade(version, success: result))
        
        return result
    }
    
}
